src/components/Question.tsx

```diff
@@ -299,18 +315,23 @@ class RenderedQuestion extends React.Component<QuestionProp> {
             if (!this.props.public_state.get("valid")) {
                 valid = false;
             }
-            const rawError = this.props.public_state.get("error");
-            let error = "";
-            if (typeof rawError === "string") {
-                error = rawError;
+
+            let error;
+            if (this.props.question.type === QuestionType.Code && this.props.public_state.get("unittestsFailed")) {
+                error = this.generateUnitTestErrorMessage(valid);
+            } else {
+                // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+                // @ts-ignore
```

Comment:
It's always a string right now, but there's no way to legally assert that unless you add a type check.

Since we're setting the error variable based on this, another check will need to be added to handle the case where the error variable was not set. This never happens though, so it would be a bunch of fun error handling for an impossibility.

The larger issue here is that state is a map with no hard defined rules on type. The same is true of many things around the codebase, such as setting the form ID, which has to be asserted to being a string.

One thing that could alleviate this would be defining an interface for state, but even then you're going to have to handle a few null checks which we know are impossible right now.

Reply:
This took a lot more work than expected, but it's done now. 09f25bed